Add a TSL2561 driver and example usage

bschwind opened the pull request:
This pull request adds support for the [TSL2561 Light Sensor](https://cdn-shop.adafruit.com/datasheets/TSL2561.pdf)

It uses the I2C module from the `extras` directory of this project.

This driver supports setting an integration time from the presets of (13ms, 101ms, and 402ms), as well as a 1X and 16X gain. The driver uses `vTaskDelay` when the lux value is requested from the sensor so it has time to integrate the sensors and get a reading. The sensor remains in a "powered down" mode when not in use.

Interrupts and custom integration times are not supported in this PR.

sheinz commented on extras/tsl2561/tsl2561.h:
```diff
@@ -0,0 +1,125 @@
+#include <stdbool.h>
+
+// I2C Addresses
+#define TSL2561_I2C_ADDR_VCC 0x49
```
All hardware related defines can be moved to .c file as it is not part of the driver's interface.

bschwind replied:
I moved most of the defines into the .c file, but there are a few I left in the header which you can use when calling particular TSL2561 functions, such as when choosing an I2C address or an integration time constant.

UncleRus replied:
I tnink enums would be better:
```C
typedef  enum {
    TSL2561_GAIN_1X = 0x00,
    TSL2561_GAIN_16X = 0x10
} tsl2561_gain_t;

...

void tsl2561_set_gain(tsl2561_t *device, tsl2561_gain_t gain);
```

sheinz commented on extras/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,287 @@
+#include "task.h"
+#include "tsl2561.h"
+
+bool write_register(uint8_t i2c_addr, uint8_t reg, uint8_t value)
```
All internal functions could be made static.

bschwind replied:
Fixed!

sheinz commented on extras/tsl2561/tsl2561.c:
```diff
@@ -0,0 +1,287 @@
+
+            break;
+        default:
+            printf("Invalid package type in CalculateLux\n");
```
In this case the function will still return true.

bschwind replied:
Good catch!

sheinz commented:
Thanks @bschwind for the contribution!

This looks good apart from a few minor comments.

bschwind commented:
@sheinz Thanks for the review!

I addressed your comments and updated the PR. When it's ready to merge I'll squash it back down to one commit, if that's preferred.

UncleRus commented on extras/tsl2561/tsl2561.h:
```diff
@@ -0,0 +1,39 @@
+#define TSL2561_GAIN_1X 0x00
+#define TSL2561_GAIN_16X 0x10
+
+typedef struct {
```
It's hard to use in cpp without 
```C
#ifdef __cplusplus
extern "C" {
#endif
```

bschwind replied:
To be clear, should the final code end up like this?

```
#ifdef  __cplusplus
extern "C" {
#endif

typedef struct {
    tsl2561_i2c_addr_t i2c_addr;
    uint8_t integration_time;
    uint8_t gain;
    uint8_t package_type;
} tsl2561_t;

#ifdef  __cplusplus
}
#endif
```

Or should the `extern "C"` block cover all symbols in tsl2561.h?

resetnow replied:
It should cover all symbols (so C++ compiler knows no name mangling should be used).

bschwind commented:
@sheinz @UncleRus @resetnow, Thanks for your reviews and suggestions
